### ISSUE TITLE
Add support for Celsius_to_Fahrenheit and Fahrenheit_to_Celsius scalar functions

### ISSUE THREAD

**Issue by archisman-panigrahi** [sharkdp/numbat#184]
I totally agree (with https://github.com/sharkdp/insect/issues/68) that Celsius and Fahrenheit are problematic units to work with, and it's best not to support any arithmetic with them.

However, Celsius and Fahrenheit are among the most commonly converted units (to be frank, after moving to the US, this is the most frequent unit conversion I mentally do everyday, using the (approximate) formula c ≈ (f-30)/2)

---
I propose we include two functions (whose input and output are both scalar), `Celsius_to_Fahrenheit` and `Fahrenheit_to_Celsius`, which will work like the following.

```
>>> Celsius_to_Fahrenheit(c)
    #Alternative 1, print the value of (32 + 9*c/5)
    #Alternative 2, print the string "c°C = (32 + 9*c/5)°F"

>>> Fahrenheit_to_Celsius(f)
    #Alternative 1, print the value of 5*(f-32)/9
    #Alternative 2, print the string "f°F = 5*(f-32)/9 °C"
```

**Comment by sharkdp** [comment]
> I totally agree (with [sharkdp/insect#68](https://github.com/sharkdp/insect/issues/68)) that Celsius and Fahrenheit are problematic units to work with, and it's best not to support any arithmetic with them.

Maybe we should revisit this and add (proper, limited) support for dealing with °C and °F directly.



> I propose we include two functions (whose input and output are both scalar), `Celsius_to_Fahrenheit` and `Fahrenheit_to_Celsius`, which will work like the following.

Note that you can do this today. Just add the following to your [`init.nbt`](https://numbat.dev/doc/cli-customization.html) file:
```
fn celsius_to_fahrenheit(t_c: Scalar) -> str =
  "{t_c}°C = {32 + 9 t_c/5}°F"

fn fahrenheit_to_celsius(t_f: Scalar) -> str=
  "{t_f}°F = {5 × (t_f-32)/9}°C"
```
(you might want to add `round`ing)

**Comment by sharkdp** [comment]
Another thing we can do right now is the following:

```
dimension TemperatureCelsius
unit °C: TemperatureCelsius

dimension TemperatureFahrenheit
unit °F: TemperatureFahrenheit

fn to_F(temp: TemperatureCelsius) -> TemperatureFahrenheit = (32 + 9 × (temp / °C) / 5) °F
fn to_C(temp: TemperatureFahrenheit) -> TemperatureCelsius = (5 × (temp / °F - 32) / 9) °C
```

And then use it like so:
```
>>> 30°C // to_F

      = 86 °F    [TemperatureFahrenheit]
```

Try it [here](https://numbat.dev/?q=dimension+TemperatureCelsius%0Aunit+%C2%B0C%3A+TemperatureCelsius%0A%0Adimension+TemperatureFahrenheit%0Aunit+%C2%B0F%3A+TemperatureFahrenheit%0A%0Afn+to_F%28temp%3A+TemperatureCelsius%29+-%3E+TemperatureFahrenheit+%3D+%2832+%2B+9+%C3%97+%28temp+%2F+%C2%B0C%29+%2F+5%29+%C2%B0F%0Afn+to_C%28temp%3A+TemperatureFahrenheit%29+-%3E+TemperatureCelsius+%3D+%285+%C3%97+%28temp+%2F+%C2%B0F+-+32%29+%2F+9%29+%C2%B0C)

**Comment by archisman-panigrahi** [comment]
I like [this](https://github.com/sharkdp/numbat/issues/184#issuecomment-1736806447) (i.e. `fn celsius_to_fahrenheit` approach) more than the second one.
 
If we add the units Celsius and Fahrenheit, then numbat will also be able to add, subtract and divide them, all of which are unphysical operations.

> Note that you can do this today. Just add the following to your [init.nbt](https://numbat.dev/doc/cli-customization.html) file:

While I know how to do this, I am suggesting we add these functions to the default interface, so that [Linux users looking for an unit converter](https://askubuntu.com/q/1484697/124466) will not have to write code to do these conversions, and we can recommend numbat to them.

**Comment by mhmdanas** [comment]
@sharkdp I like your suggestion.

> If we add the units Celsius and Fahrenheit, then numbat will also be able to add, subtract and divide them, all of which are unphysical operations.

That shouldn't be a problem if each has its own dimension, right? Because then you can't combine the different temperature units without explicitly converting them.

**Comment by sharkdp** [comment]
> > If we add the units Celsius and Fahrenheit, then numbat will also be able to add, subtract and divide them, all of which are unphysical operations.
> 
> That shouldn't be a problem if each has its own dimension, right? Because then you can't combine the different temperature units without explicitly converting them.

No, @archisman-panigrahi is right. It is still a problem. Adding two values in °C usually doesn't make any sense (unless one of them is an *absolute* value and the other one a *relative* value). It's kind of similar to DateTimes and durations.

I think we probably need to add proper support in the language for units with additive offsets. A first step would be to add support for units that you can't do anything with... except convert to and from.

**Comment by domsl** [comment]
Hi, first of all, I love numbat, I used to do my quick calculations in python, but numbat is much nicer :)
I just stumbled across this issue, It took me a minute to find the functions `celsius()` and `from_celsius()`

I think it always make sense to fall back to K, but most people I know calculate in celsius, as it is often okay to do if its linear (as a difference of 1K is 1 °C):

![Image](https://github.com/user-attachments/assets/95945576-d67a-4534-99a7-fb3a1edda979)

So one option could be to always implicitly convert °C to K every time for the calculation and then show the output in °C again?

I know you could do that with `h*2m²*(from_celsius(100)-from_celsius(99))` instead of `h*2m²*(100C-99C)`, but it would be much nicer to write/read? :)

**Comment by Goju-Ryu** [comment]
> So one option could be to always implicitly convert °C to K every time for the calculation and then show the output in °C again?

I have had similar thoughts on the issue and agree it would be nice to be able to use celsius for many physics equations. Most equations in my high school physics back when, used celsius and I would have loved to use numbat for those classes if it was around.  
I have been idly thinking about using the same mechanism as in [this comment](https://github.com/sharkdp/numbat/issues/481#issuecomment-2661427325) to achieve it, but haven't tried it out yet.

**Comment by haborite** [comment]
Hi, thank you for creating such a wonderful language.

Regarding the issue — as a Julia user, I think the approach taken by [Unitful.jl](https://painterqubits.github.io/Unitful.jl/stable/temperature/) seems to work well. 

With this approach, all the following procedure is valid:

```julia
using Unitful

h1 = 10u"m"
h2 = 110u"m"

T1 = 30.0u"°C"
T2 = 29.4u"°C"

ΔT = T2 - T1
Δh = h2 - h1

gradient = ΔT / Δh

T0 = 40u"°C"
Δh = 100u"m"

ΔT = gradient * Δh
T_new = T0 + ΔT

println("T: ", T_new)
# 312.54999999999995 K
```

Whereas the following is not allowed:

```
T1 + T2
# LoadError: AffineError: an invalid operation was attempted with affine quantities: 
```

This may already be something that's starting to be implemented in #462, but I thought it might be helpful to share a user's impression.

**Comment by sharkdp** [comment]
°C and °F are now supported, see https://numbat.dev/docs/basics/conversions/#temperature-conversions for details. And see #825 if you're interested in the limitations and the implementation.

**Comment by haborite** [comment]
I'm thrilled to hear this feature has been implemented! Thank you for your work.
